File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/34_find-first-and-last-position-of-element-in-sorted-array/solutions/cpp/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result(2, -1);
        int left = 0, right = nums.size() - 1;
        
        // Find the first occurrence of target
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                // Move leftwards to find the start
                if (mid == 0 || nums[mid - 1] != target) {
                    result[0] = mid;
                    break;
                } else {
                    right = mid - 1;
                }
            }
        }
        
        left = 0;
        right = nums.size() - 1;
        
        // Find the last occurrence of target
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                // Move rightwards to find the end
                if (mid == nums.size() - 1 || nums[mid + 1] != target) {
                    result[1] = mid;
                    break;
                } else {
                    left = mid + 1;
                }
            }
        }
        
        return result;
    }
};

File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/34_find-first-and-last-position-of-element-in-sorted-array/solutions/python3/solution.py
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarySearchLeft():
            left, right = 0, len(nums) - 1
            while left <= right:
                mid = left + (right - left) // 2
                if nums[mid] < target:
                    left = mid + 1
                elif nums[mid] > target:
                    right = mid - 1
                else:
                    if mid == 0 or nums[mid - 1] != target:
                        return mid
                    right = mid - 1
            return -1

        def binarySearchRight():
            left, right = 0, len(nums) - 1
            while left <= right:
                mid = left + (right - left) // 2
                if nums[mid] < target:
                    left = mid + 1
                elif nums[mid] > target:
                    right = mid - 1
                else:
                    if mid == len(nums) - 1 or nums[mid + 1] != target:
                        return mid
                    left = mid + 1
            return -1

        left_index = binarySearchLeft()
        right_index = binarySearchRight()
        return [left_index, right_index] if left_index != -1 else [-1, -1]
